modules/mono/build_scripts/build_assemblies.py

```diff
@@ -353,6 +368,77 @@ def build_all(msbuild_tool, module_dir, output_dir, godot_platform, dev_debug, p
     return 0
 
 
+def run_quiet(args):
+    try:
+        subprocess.check_output(args)
+    except Exception as e:
+        pass
+
+
+def make_assemblies(target, source, env):
+    exit_code = start_build(
+        godot_output_dir="bin/",
+        godot_platform=env["platform"],
+        dev_debug=env["dev_build"],
+        mono_prefix="",
+        precision=env["precision"],
+        push_nupkgs_local=False,
+        clear_nuget_cache=True,
+    )
+
+    if exit_code != 0:
+        return exit_code
+
+    nupkgs_dir = os.path.abspath("bin/GodotSharp/Tools/nupkgs")
+    tools = find_any_msbuild_tool("")
+
+    # configure a default nuget source pointing to the bin/GodotSharp/Tools/nupkgs dir, so packages can be picked up automatically
```

Comment:
I understand, it's a reasonable hesitation. However, without this, C# projects just won't work, which means it is a mandatory step to the whole pipeline - you have to configure a nuget source in order to pick up the packages. Having this opt-in means 99% of users having to run this manually anyways, which defeats the purpose somewhat.

I agree that it's happening too quietly, and with no way to opt-out of it if you have your own custom setup. I don't think it should be opt-in - we should be aiming for the 80/20 rule in user experience (making the 80% most common workflows easier) - but it should be controllable by a flag, so you can disable the behaviour in your own setup and not disrupt thing for very advanced users. I'll add a thing for that.

Reply:
We've been bitten in the past by making changes to user configuration. In 4.0 we had a `GodotNuGetFallbackFolder` and changed the user configuration so NuGet would use it. The problem is that, if that folder is removed, then NuGet doesn't work at all. This means we essentially broke all their .NET projects (not just Godot projects), because they could no longer restore packages.

Configuring a NuGet source is a step that only needs to be done once though, so I don't think it's that much to ask. It's the same thing as installing the .NET SDK or setting up other prerequisites before building Godot, and we don't automate those either.

Reply:
I've actually experienced this issue last time I tested this PR (and the issue of course happens if following the current manual steps, but it's more explicit to the user).

I compiled Godot with Mono support, it generated the local nuget source and everything worked great out of the box :+1: 

Then some days later I ran `git clean -fxd` as I often do, it removed the folder from `bin`, and next time I tried to used an official build of Godot, my .NET setup was broken. I had to ask on chat and be told to check whether my `~/.nuget/NuGet/NuGet.config` was referring to something missing.

Is there no way to add a `packageSource` in that config that wouldn't break dotnet if it's missing? Like with a conditional so it's only added if the folder exists?

Reply:
> Is there no way to add a `packageSource` in that config that wouldn't break dotnet if it's missing? Like with a conditional so it's only added if the folder exists?

I don't think so, I remember looking at a way to do this in the context of `GodotNuGetFallbackFolder` but I couldn't find anything.

Reply:
Hm :\

I don't know if it's easy to do with scons, but maybe we could make the option output information after the build (the first time the package source needs to be added only) that tells the user about what was done, and how to remove the package source if they need?

I feel like just making the behavior opt-in wouldn't help much because if it's a SCons options, most users won't know what it does and how to undo it anyway.

Another option could be to install the packageSource stuff in the user's NuGet folder (`~/.nuget`), so that they won't easily delete it by mistake (and I guess it's not too harmful to their other non-Godot stuff, it just has Godot related packages).

Reply:
You're absolutely right, I also hit this right after commenting, completely forgot that nuget is the worst. Urgh.
There are three "ways" of avoiding the system nuget failing miserably when a local source path doesn't exist:

- Nuget supports a `--ignore-failed-sources` flag on restore to not fail in these situations. You have to pass it in manually, though, so that doesn't help anything, as users (and VS) won't pass this flag in normally.

- Pushing local build packages to a location that is known to always exist: I was trying to avoid pushing packages to another place in the filesystem and instead consume them directly from the bin folder, because that would avoid stale/forgotten local build packages being picked up accidentally, but that's indeed a safe option.

- Generating nuget.config files: The only way to have nuget sources that don't affect every project is to have a `nuget.config` file next to the csproj, with the local source, so only that project is affected if a source goes away. We could generate this config file in the project folder when a project is opened with a local build of godot-dotnet. Since the csproj in these cases would already referencing a `-dev` version of the Sdk, it would fail to restore with non-local builds of godot, a failure to restore if the local source path is missing would make it obvious what the failure is, without being confusing to other .net projects.

Reply:
I don't think NuGet is the _worst_, it's just probably not designed for this use case. I think it's weird for a project to add a local source, if you let the user choose the source then it can be whatever they want (it could be a remote source).

Since your intention is to make this "just work", I agree that the first option is not a solution.

The second option is what @akien-mga suggested. I don't really like it, we're still modifying the user's configuration and now we're storing packages somewhere in their computer that they may not be aware of. It feels rude.

I like the third option, but we have to keep in mind that the user may already have a `nuget.config` file in their project. Not sure how difficult it would be to manage this situation, but it feels like it'd be overkill to implement all this if it's not the most common case.

My opinion is that we're overcomplicating this. Setting up a local source is a step you only have to do **once**, are there really users complaining about this step or is it a theoretical problem that we're trying to solve here?